library/src/scripts/flyouts/NewPostMenu.tsx

```diff
@@ -59,47 +71,288 @@ function ActionItem({ item, style }: { item: IAddPost; style: ITransition }) {
                     {contents}
                 </LinkAsButton>
             )}
-        </div>
+        </animated.li>
     );
 }
 
+interface IState {
+    open: boolean;
+    buttonFocus: boolean;
+    focusedItem?: number;
+}
+
+const initialState: IState = {
+    open: false,
+    buttonFocus: false,
+    focusedItem: undefined,
+};
+
+const reducer = (state, action) => {
+    switch (action.type) {
+        case "toggle_open":
+            return { ...state, open: !state.open };
+        case "set_open":
+            return { ...state, open: action.open };
+        case "set_button_focus":
+            return { ...state, buttonFocus: action.focus };
+        case "set_focused_item":
+            return { ...state, focusedItem: action.item };
+        default:
+            throw new Error();
+    }
+};
+
 export default function NewPostMenu(props: { items: IAddPost[] }) {
-    const [open, setOpen] = useState(false);
-    const toggle = () => setOpen(!open);
-    const onClickBackground = () => {
-        if (open) {
-            setOpen(false);
+    const classes = newPostMenuClasses();
+    const vars = newPostMenuVariables();
+
+    let { items } = props;
+
+    const buttonRef = useRef<HTMLButtonElement>(null);
+    const backgroundRef = useRef<HTMLElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const itemsRef = useRef<HTMLLIElement>(null);
+    const accessMenuRef = useRef<HTMLUListElement>(null);
+
+    const [state, dispatch] = useReducer(reducer, initialState);
+
+    const onClickBackground = (event: React.MouseEvent) => {
+        event.stopPropagation();
+        if (state.open) {
+            dispatch({ type: "set_button_focus", focus: true });
+            dispatch({ type: "set_open", open: false });
+            dispatch({ type: "set_focused_item", item: undefined });
         }
     };
 
-    const classes = newPostMenuClasses();
-    const { items } = props;
+    const onBgKeyDown = (event: React.KeyboardEvent<any>) => {
+        switch (event.key) {
+            case "Escape":
+                if (state.open) {
+                    event.preventDefault();
+                    event.stopPropagation();
+                    dispatch({ type: "set_open", open: false });
+                    dispatch({ type: "set_button_focus", focus: true });
+                    dispatch({ type: "set_focused_item", item: undefined });
+                }
+                break;
+            default:
+                break;
+        }
+    };
+
+    const onMenuItemKeyDown = (event: React.KeyboardEvent<any>) => {
+        const keys = ["Escape", "Home", "End", "ArrowUp", "ArrowDown", "Enter"];
+
+        // Only handling keys in these situations so preventDefault and stopPropagation in these
+        if (
+            keys.includes(event.key) ||
```

Comment:
This does not work, you also check for the open state. You're blocking the user from tabbing here, that's bad. I told you these are probably not needed and cause problems. If they are needed, you need to put them in your case, in the "if" statement.

Reply:
I just deleted those according to you.